[BUGFIX release] Correct types for Ember Arrays

wagenet opened the pull request:
This PR does a couple of things to improve array types:
* More fully unifies internal array types with external preview types.
* Corrects NativeArray types to not override methods from the true native JS array.

Fixes #20230

wagenet committed:
Improve types for Ember Arrays

chriskrycho commented:
Top-level request (I'll review further later): can you add an explanation to the PR so we can refer to it later to understand *why* these specific changes are correct?

chriskrycho commented on packages/@ember/-internals/utils/types.ts:
```diff
@@ -1,7 +1,15 @@
 export type AnyFn = (...args: any[]) => any;
 
+export type MethodsOf<O> = {
+  [K in keyof O]: O[K] extends AnyFn ? O[K] : never;
+};
```
This doesn't actually work, unfortunately. You need something like the final `keyof` as used in `MethodNamesOf` below to make it properly strip out the bits which are actually `never`. It turns out the easiest way to get `MethodsOf` is to reuse `MethodNamesOf` in conjunction with `Pick`:

```ts
type MethodsOf<O> = Pick<O, MethodNamesOf<O>>;
```

[playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggdiAYnKBeKAKAdDghgJwHMBnALilwQG0BdASjQD4KEBuAKHYgA8wB7fMCihIUALIRgACz4ATYgHkAZgB4FzdAG92UKFQDSUAJYoA1hBB8lUBTXIKDNKD2AQ482AmRQA-DcdQ5HAQAG4Q+BwAvhzsItAAQpKu+BLScoqq6mhQAApGAMamagA04pIysgByuAC2EBlqjIwxPPyCwuDQqRXVdQ0AKhpQ2rp6hiZQ5pbW-XZQ-QEubh7wSCh+hkGh4RxQkVRTVvM0MQD0p1AAVgCuxELEfFCW11D5lFDEENAyAO7GQj8BKZiFAfkZpMIpNAeLUwAAbepQABGEDhfB+7FkEHycII0BxuGIILEuBAKJUg2GOlGIVwcOuEB85H6u1GNVwYBUAFVGBglHByBhgMyGKhmFy6OQSWSINzmuxIpw4lAABrZbpyXr1ZQqaXk65wUxwdFwJocZUATXV5XSOr1soNRpNZtinSgAC1solgMkNfI7aT9YbjT9Tc0gA)

wagenet commented on types/preview/@ember/array/mutable.d.ts:
```diff
@@ -15,7 +15,7 @@ declare module '@ember/array/mutable' {
      * __Required.__ You must implement this method to apply this mixin.
      */
-    replace(idx: number, amt: number, objects: T[]): this;
+    replace(idx: number, amt: number, objects: T[]): void;
```
In some cases we do return `this`, but the docs state that we do not return.

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -515,7 +519,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  getEach<K extends string>(key: K): NativeArray<Value<T, K>>;
+  getEach<K extends keyof T>(key: K): NativeArray<T[K]>;
+  getEach<K extends CompoundPropertyKey<T>>(key: K): NativeArray<unknown>;
```
Instead of introducing additional type shenanigans here, since it ultimately just becomes `NativeArray<unknown>`, we can just use `string`. (This is how we do it with the overloads in Ember and Ember Data).

```suggestion
  getEach(key: string): NativeArray<unknown>;
```

wagenet replied:
Resolved.

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -20,12 +20,15 @@ import MutableEnumerable from '@ember/enumerable/mutable';
 
-type Value<T, K extends string> = K extends keyof T ? T[K] : unknown;
+export type EmberArrayLike<T> = EmberArray<T> | NativeArray<T>;
+
+// We don't currently attempt to fully type-check compound keys
+type CompoundPropertyKey<T> = `${keyof T & string}.${string}`;
```
Then we don't need this!

```suggestion
```

wagenet replied:
Resolved

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -746,10 +752,6 @@ interface EmberArray<T> extends Enumerable {
-  find<S extends T, Target = void>(
-    predicate: (this: void, value: T, index: number, obj: T[]) => value is S,
-    thisArg?: Target
-  ): S | undefined;
```
Just moved this.

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -1061,10 +1070,10 @@ interface EmberArray<T> extends Enumerable {
+  invoke<M extends MethodNamesOf<T>>(
+    methodName: M,
+    ...args: MethodParams<T, M>
+  ): NativeArray<MethodReturns<T, M>>;
```
The public preview types were better here.

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -1086,7 +1095,7 @@ interface EmberArray<T> extends Enumerable {
     @public
   */
-  compact(): NativeArray<Exclude<T, null>>;
+  compact(): NativeArray<NonNullable<T>>;
```
The public preview types were better here.

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -1176,6 +1185,7 @@ interface EmberArray<T> extends Enumerable {
     @public
   */
   uniqBy(key: string): NativeArray<T>;
+  uniqBy(callback: (value: T) => unknown): NativeArray<T>;
```
This was missing before.

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -1632,7 +1642,7 @@ interface MutableArray<T> extends EmberArray<T>, MutableEnumerable {
     @public
   */
-  unshiftObject(object: T): this;
+  unshiftObject(object: T): T;
```
This was incorrect before.

wagenet commented on packages/@ember/array/index.ts:
```diff
@@ -1892,6 +1902,167 @@ const MutableArray = Mixin.create(EmberArray, MutableEnumerable, {
+
+// This is the same as MutableArray, but removes the actual native methods that exist on Array.prototype.
+interface MutableArrayWithoutNative<T>
+  extends Omit<MutableArray<T>, IGNORED_MUTABLE_ARRAY_METHODS | RETURN_SELF_ARRAY_METHODS> {
```
NativeArray is MutableArray with all actual native methods removed. Unfortunately, because of how Omit works, all methods that return `this` will end up not returning the correct `this`, so we also have to redefine those methods here.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -37,9 +40,9 @@ expectTypeOf(arr.firstObject).toEqualTypeOf<Foo | undefined>();
-expectTypeOf(arr.slice(1, 2)).toEqualTypeOf<NativeArray<Foo>>();
+expectTypeOf(arr.slice()).toEqualTypeOf<Foo[]>();
+expectTypeOf(arr.slice(1)).toEqualTypeOf<Foo[]>();
+expectTypeOf(arr.slice(1, 2)).toEqualTypeOf<Foo[]>();
```
We're actually hitting the native slice.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -49,7 +52,7 @@ expectTypeOf(arr.lastIndexOf(new Foo())).toEqualTypeOf<number>();
 arr.lastIndexOf('invalid');
 
-expectTypeOf(arr.forEach((item: Foo) => String(item))).toEqualTypeOf(arr);
+expectTypeOf(arr.forEach((item: Foo) => String(item))).toEqualTypeOf<void>();
```
Native forEach returns void.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -58,43 +61,44 @@ arr.forEach((item, index, arr) => {
   expectTypeOf(arr).toEqualTypeOf(arr);
 });
 arr.forEach(function (_item) {
-  expectTypeOf(this).toEqualTypeOf(target);
```
Native methods behave differently here.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -58,43 +61,44 @@ arr.forEach((item, index, arr) => {
 expectTypeOf(arr.getEach('bar')).toEqualTypeOf<NativeArray<number>>();
-expectTypeOf(arr.getEach('missing')).toEqualTypeOf<NativeArray<unknown>>();
+// @ts-expect-error property is unknown
+arr.getEach('missing');
```
We updated types to not allow this case.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -58,43 +61,44 @@ arr.forEach((item, index, arr) => {
-// Can set unknown property
-expectTypeOf(arr.setEach('missing', 'anything')).toEqualTypeOf(arr);
+// @ts-expect-error property is unknown
+arr.setEach('missing', 'anything');
```
We updated types to not allow this.

wagenet commented on packages/@ember/array/type-tests/index.ts:
```diff
@@ -166,16 +172,14 @@ let reduced = arr.reduce((summation, item, index, arr) => {
 expectTypeOf(reduced).toEqualTypeOf<number>();
-// NOTE: This doesn't match native behavior and is a bit weird
-expectTypeOf(arr.reduce((summation, _item) => summation)).toEqualTypeOf<unknown>();
+expectTypeOf(arr.reduce((summation, _item) => summation)).toEqualTypeOf<Foo>();
```
This does match native types now.

wagenet commented on packages/@ember/array/type-tests/mutable.ts:
```diff
@@ -1,63 +1,67 @@
 
-class Foo {}
+class Foo {
+  constructor(public name: string) {}
```
This needed a property for the checks below to fully work.

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -593,7 +598,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  mapBy<K extends string>(key: K): NativeArray<Value<T, K>>;
+  mapBy<K extends keyof T>(key: K): NativeArray<T[K]>;
+  mapBy<K extends CompoundPropertyKey<T>>(key: K): NativeArray<unknown>;
```
```suggestion
  mapBy(key: string): NativeArray<unknown>;
```

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -824,7 +830,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  findBy<K extends string>(key: K, value?: Value<T, K>): T | undefined;
+  findBy<K extends keyof T>(key: K, value?: T[K]): T | undefined;
+  findBy<K extends CompoundPropertyKey<T>>(key: K, value?: unknown): T | undefined;
```
```suggestion
  findBy(key: string, value?: unknown): T | undefined;
```

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -904,7 +911,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  isEvery<K extends string>(key: K, value?: Value<T, K>): boolean;
+  isEvery<K extends keyof T>(key: K, value?: T[K]): boolean;
+  isEvery<K extends CompoundPropertyKey<T>>(key: K, value?: unknown): boolean;
```
```suggestion
  isEvery(key: string, value?: unknown): boolean;
```

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -972,7 +980,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  isAny<K extends string>(key: K, value?: Value<T, K>): boolean;
+  isAny<K extends keyof T>(key: K, value?: T[K]): boolean;
+  isAny<K extends CompoundPropertyKey<T>>(key: K, value?: unknown): boolean;
```
```suggestion
  isAny(key: string, value?: unknown): boolean;
```

wagenet commented on packages/@ember/debug/container-debug-adapter.ts:
```diff
@@ -88,9 +88,9 @@ export default class ContainerDebugAdapter extends EmberObject {
-    let types = emberA<string>();
+  catalogEntriesByType(type: string): string[] {
+    let namespaces = Namespace.NAMESPACES;
+    let types: string[] = [];
```
This is just documented as returning an array.

wagenet commented on packages/@ember/debug/data-adapter.ts:
```diff
@@ -544,22 +544,22 @@ export default class DataAdapter<T> extends EmberObject {
 
     // New adapters return strings instead of classes.
-    let klassTypes = emberA(stringTypes).map((name) => {
+    let klassTypes = stringTypes.map((name) => {
```
These uses of `emberA` were unnecessary since `map` and `filter` are native methods.

chriskrycho replied:
On the one hand, this is *definitely* a change I'm excited to be able to make; on the other, my note here is the same as above. Let's do it in a separate PR so it's easy to trace back to if for some reason it ends up being an issue for folks.

wagenet commented on packages/@ember/debug/data-adapter.ts:
```diff
@@ -544,22 +544,22 @@ export default class DataAdapter<T> extends EmberObject {
     @return {Array} Array of model types.
   */
-  getModelTypes(): NativeArray<{ klass: unknown; name: string }> {
+  getModelTypes(): Array<{ klass: unknown; name: string }> {
```
This is just documented as returning an array.

wagenet commented on type-tests/preview/@ember/array-test/array.ts:
```diff
@@ -26,13 +27,11 @@ expectTypeOf(people.objectAt(0)).toEqualTypeOf<Person | undefined>();
 
 expectTypeOf(people.filterBy('isHappy')).toMatchTypeOf<Person[]>();
-expectTypeOf(people.filterBy('isHappy')).toMatchTypeOf<MutableArray<Person>>();
+expectTypeOf(people.filterBy('isHappy')).toMatchTypeOf<NativeArray<Person>>();
```
NativeArray !== MutableArray

wagenet commented on type-tests/preview/ember/ember-module-tests.ts:
```diff
@@ -153,7 +153,7 @@ expectTypeOf(Ember.Application.create()).toEqualTypeOf<Ember.Application>();
 expectTypeOf(Ember.ApplicationInstance.create()).toEqualTypeOf<Ember.ApplicationInstance>();
 // Ember.Array
-const a1: Ember.Array<string> = Ember.A([]);
+const a1: Ember.NativeArray<string> = Ember.A([]);
```
NativeArray !== EmberArray

wagenet commented on types/preview/@ember/array/-private/native-array.d.ts:
```diff
@@ -1,22 +1,180 @@
+   * The final definition of NativeArray removes all native methods. This is the list of removed methods
+   * when run in Chrome 106.
+   */
+  type IGNORED_MUTABLE_ARRAY_METHODS =
```
The below should be the same as internal.

wagenet commented on types/preview/@ember/array/index.d.ts:
```diff
@@ -4,6 +4,9 @@ declare module '@ember/array' {
   import NativeArray from '@ember/array/-private/native-array';
 
+  // We don't currently attempt to fully type-check compound keys
+  type CompoundPropertyKey<T> = `${keyof T & string}.${string}`;
```
Bring these more in line with internal types.

wagenet commented on types/preview/@ember/array/mutable.d.ts:
```diff
@@ -45,12 +45,12 @@ declare module '@ember/array/mutable' {
      * it is KVO-compliant.
      */
-    popObject(): T;
+    popObject(): T | null | undefined;
```
These types were incorrect before.

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -515,7 +519,8 @@ interface EmberArray<T> extends Enumerable {
   */
-  getEach<K extends string>(key: K): NativeArray<Value<T, K>>;
+  getEach<K extends keyof T>(key: K): NativeArray<T[K]>;
+  getEach<K extends CompoundPropertyKey<T>>(key: K): NativeArray<unknown>;
```
Ah, so this catches the case where we have `get(anArray, 'lol-nope')` and makes it not type check. In general, we haven’t tried to do that: people *do* (and from a public API can :upside_down:) stash unknown stuff on instances with `set` and then later `get` it, which should be "safe" (if not recommended).

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -792,6 +790,10 @@ interface EmberArray<T> extends Enumerable {
     @return {Object} Found item or `undefined`.
     @public
   */
+  find<S extends T, Target = void>(
```
Note: conventionally/idiomatically it's `U` in a case like this (since T → U in the alphabet). Not a thing I care about changing, to be sure, just worth noting!

wagenet replied:
I think I just copied this from somewhere else :)

chriskrycho commented on packages/@ember/array/index.ts:
```diff
@@ -1892,6 +1898,167 @@ const MutableArray = Mixin.create(EmberArray, MutableEnumerable, {
+
+type AnyArray<T> = EmberArray<T> | Array<T>;
+
+/**
+ * The final definition of NativeArray removes all native methods. This is the list of removed methods
+ * when run in Chrome 106.
+ */
```
Note for posterity/future readers looking at this PR historically—we definitely need this, but we need even more to just get the heck rid of it, since this could in principle change dynamically over time.

chriskrycho commented on packages/@ember/array/type-tests/mutable.ts:
```diff
@@ -1,63 +1,67 @@
 
 expectTypeOf(arr.replace(1, 1, [foo])).toEqualTypeOf<void>();
-// TODO: Why doesn't this fail?
+// @ts-expect-error invalid item
```
It's nice that all of these fail now the way we want them to. 👍🏼

wagenet replied:
I think we just needed a better test item :)

chriskrycho commented on packages/@ember/debug/container-debug-adapter.ts:
```diff
@@ -88,9 +86,9 @@ export default class ContainerDebugAdapter extends EmberObject {
-  catalogEntriesByType(type: string): NativeArray<string> {
-    let namespaces = emberA(Namespace.NAMESPACES);
-    let types = emberA<string>();
+  catalogEntriesByType(type: string): string[] {
+    let namespaces = Namespace.NAMESPACES;
+    let types: string[] = [];
```
Hmm. Documentation is very apt to be incorrect in this space, though; folks who wrote the docs likely were not thinking about the differences (because the differences are subtle!). Let's avoid changing the behavior here—we can always do so in a later PR, but keeping types changes separate from behavior changes will help us avoid snafus.

chriskrycho commented on packages/@ember/debug/data-adapter.ts:
```diff
@@ -571,8 +571,8 @@ export default class DataAdapter<T> extends EmberObject {
   _getObjectsOnNamespaces() {
-    let namespaces = emberA(Namespace.NAMESPACES);
-    let types = emberA<string>();
+    let namespaces = Namespace.NAMESPACES;
+    let types: string[] = [];
```
Same here, presumably.

chriskrycho reviewed (changes requested):
Couple of tweaks needed, but this is a really nice step forward. 💙

chriskrycho committed:
Rework `MethodsOf` type utilities for robustness

1. Using the keyof constraint feature introduced in 4.1, avoid ever
   introducing types with properties whose type is `never` for the
   non-method fields.

2. Handle recursively applying this constraint to nested methods for
   the `EmberMethodsOf` type, which has additional complexity to handle
   Ember's string-based dispatch. Combined with (1), this actually
   *simplifies* the types, eliminating one layer of conditional types.
